File: feb11/backend/utils/merge_utils.py
def deep_merge(dst: dict, src: dict) -> dict:
    """
    Canonical deep merge: merge src into dst.
    - Dicts: recursively merged.
    - Lists: merged element-by-element by index, preserving extras in dst.
    - Scalars: src overwrites dst.
    """
    if not isinstance(src, dict):
        return src
    if not isinstance(dst, dict):
        return src

    result = dst.copy()
    for key, src_val in src.items():
        if key not in result:
            result[key] = src_val
        elif isinstance(src_val, dict) and isinstance(result[key], dict):
            result[key] = deep_merge(result[key], src_val)
        elif isinstance(src_val, list) and isinstance(result[key], list):
            merged_list = result[key].copy()
            for i, src_elem in enumerate(src_val):
                if i < len(merged_list):
                    if isinstance(src_elem, dict) and isinstance(merged_list[i], dict):
                        merged_list[i] = deep_merge(merged_list[i], src_elem)
                    else:
                        merged_list[i] = src_elem
                else:
                    merged_list.append(src_elem)
            result[key] = merged_list
        else:
            result[key] = src_val
    return result

File: feb11/frontend/src/lib/mergeUtils.js
/**
 * Canonical deep merge: merge src into dst.
 * - Objects: recursively merged.
 * - Arrays: merged element-by-element by index, preserving extras in dst.
 * - Scalars: src overwrites dst.
 */
export function deepMergeObjects(dst, src) {
  if (!src || typeof src !== 'object') return dst;
  if (!dst || typeof dst !== 'object') return src;

  const result = Array.isArray(dst) ? [...dst] : { ...dst };

  if (Array.isArray(src) && Array.isArray(dst)) {
    for (let i = 0; i < src.length; i++) {
      if (i < result.length) {
        if (src[i] && typeof src[i] === 'object' && !Array.isArray(src[i]) &&
            result[i] && typeof result[i] === 'object' && !Array.isArray(result[i])) {
          result[i] = deepMergeObjects(result[i], src[i]);
        } else {
          result[i] = src[i];
        }
      } else {
        result.push(src[i]);
      }
    }
    return result;
  }

  for (const [key, value] of Object.entries(src)) {
    if (value && typeof value === 'object' && !Array.isArray(value) &&
        result[key] && typeof result[key] === 'object' && !Array.isArray(result[key])) {
      result[key] = deepMergeObjects(result[key], value);
    } else if (Array.isArray(value) && Array.isArray(result[key])) {
      result[key] = deepMergeObjects(result[key], value);
    } else {
      result[key] = value;
    }
  }
  return result;
}
